### PR TITLE
meson: Clean up build and infra

### LINKED ISSUE #58831: Meson overrides compiler, ignoring stdenv
When I build a meson package (e.g. gnome-builder) with changed `stdenv` (e.g to `gcc8Stdenv`) it still builds with the compiler used by Meson:

https://github.com/NixOS/nixpkgs/blob/7c337cc807f2618e5f162a647e974cffcda86122/pkgs/development/tools/build-managers/meson/default.nix#L60

cc @Ericson2314

### PR DESCRIPTION
**Pull request by Ericson2314**
###### Motivation for this change

The big thing that happened is that https://github.com/mesonbuild/meson/pull/6363 was finally merged!

Fixes #58831

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

CC @matthewbauer @jtojnar

### CODE REVIEW

**Comment by worldofpeace** [comment]
> The big thing that happened is that #6363 was finally merged!

That seems like it's unrelated

**Comment by Ericson2314** [comment]
Oops sorry! meant https://github.com/mesonbuild/meson/pull/6363

File: pkgs/stdenv/generic/make-derivation.nix [hunk 1]
```diff
@@ -252,6 +253,41 @@ in rec {
...
+              aarch64  = "aarch64";
+              armv5tel = "arm";
+              armv6l   = "arm";
+              armv7l   = "arm";
+              i686     = "x86";
+              x86_64   = "x86_64";
+            };
+            crossFile = builtins.toFile "cross-file.conf" (''
+              [properties]
+              needs_exe_wrapper = true
```
Comment (matthewbauer) [inline pkgs/stdenv/generic/make-derivation.nix:269]: I don’t think we should set this.

File: pkgs/stdenv/generic/make-derivation.nix [hunk 2]
```diff
@@ -252,6 +253,41 @@ in rec {
           ++ lib.optional (stdenv.buildPlatform.uname.system != null) "-DCMAKE_HOST_SYSTEM_NAME=${stdenv.buildPlatform.uname.system}"
           ++ lib.optional (stdenv.buildPlatform.uname.processor != null) "-DCMAKE_HOST_SYSTEM_PROCESSOR=${stdenv.buildPlatform.uname.processor}"
           ++ lib.optional (stdenv.buildPlatform.uname.release != null) "-DCMAKE_HOST_SYSTEM_VERSION=${stdenv.buildPlatform.uname.release}";
+
+          mesonFlags = if mesonFlags == null then null else let
```
[the remark this answers is not in the document]
Comment (matthewbauer) [reply pkgs/stdenv/generic/make-derivation.nix:257]: Also we should just generate this file when cross-compiling.

[the remark this answers is not in the document]
Comment (Ericson2314) [reply pkgs/stdenv/generic/make-derivation.nix:257]: how can we check without some nasty `O(n)` scan looking for magic value in `passthru`? I rather just wait.

Fair point to the `null` trick being bad.

File: pkgs/stdenv/generic/make-derivation.nix [hunk 3]
```diff
@@ -252,6 +253,41 @@ in rec {
...
+            # See https://mesonbuild.com/Reference-tables.html#cpu-families
+            cpuFamilies = {
+              aarch64  = "aarch64";
+              armv5tel = "arm";
+              armv6l   = "arm";
+              armv7l   = "arm";
+              i686     = "x86";
+              x86_64   = "x86_64";
+            };
```
[the remark this answers is not in the document]
Comment (Ericson2314) [reply pkgs/stdenv/generic/make-derivation.nix:266]: Fair, but let's do that in a follow-up? I just moved it as-is for now.

File: pkgs/stdenv/generic/make-derivation.nix (hunk 1 above)
Comment (Ericson2314) [reply pkgs/stdenv/generic/make-derivation.nix:269]: Well, it was what was there so I rather change separately, but I do agree. Didn't you make a thing just for this sort of thing?

File: pkgs/stdenv/generic/make-derivation.nix (hunk 2 above)
[the remark this answers is not in the document]
Comment (Ericson2314) [reply pkgs/stdenv/generic/make-derivation.nix:257]: Ah so CMake does do the null thing, but I made `[]` the default so if the user just uses the default meson flags things will work.

File: pkgs/stdenv/generic/make-derivation.nix [hunk 4]
```diff
@@ -252,6 +253,41 @@ in rec {
...
+              [properties]
+              needs_exe_wrapper = true
+
+              [host_machine]
+              system = '${stdenv.targetPlatform.parsed.kernel.name}'
+              cpu_family = '${cpuFamilies.${stdenv.targetPlatform.parsed.cpu.name}}'
+              cpu = '${stdenv.targetPlatform.parsed.cpu.name}'
+              endian = ${if stdenv.targetPlatform.isLittleEndian then "'little'" else "'big'"}
+
+            '' + lib.optionalString (stdenv.hasCC) (''
```
[the remark this answers is not in the document]
Comment (Ericson2314) [reply pkgs/stdenv/generic/make-derivation.nix]: Gone!

**Comment by Ericson2314** [comment]
@matthewbauer so what i ended up doing is removing most of the items from `[binaries]` as 1) meson with my tiny patch will just grab from env vars, even during cross, and 2) that avoids `hasCC` altogether.

**Review by matthewbauer — APPROVED** [review]

**Comment by Ericson2314** [comment]
Ugh I missed ofborg telling me I broke eval on staging. Fixing in a moment.

**Pull request merged**